File: Shared/ProjectorKit/Sources/ProjectorKit/WireProtocol.swift
import Foundation

// The JSON envelope carried by every WebSocket text frame between Sender and
// Receiver (see vga/Sources/ReceiverProtocol/WireProtocol.swift). One
// request -> one response, correlated by `id`, since multiple requests can
// be in flight on the same persistent connection.
//
// This app never mirrors the screen, so RequestPayload has no `offer` case
// — but ResponsePayload keeps `answer` for wire parity with Receiver, which
// can send it to other Senders on the same protocol.

public struct ReceiverRequest: Codable, Sendable, Equatable {
    public let id: UUID
    public let payload: RequestPayload

    public init(id: UUID = UUID(), payload: RequestPayload) {
        self.id = id
        self.payload = payload
    }
}

public enum RequestPayload: Sendable, Equatable {
    /// Raw share text from the sending app — a bare URL, or freeform text
    /// with a URL embedded in it (e.g. "I'm watching X on Y\nhttps://...").
    /// Receiver is responsible for figuring out which video provider, if
    /// any, this came from.
    case video(payload: String)
    case control(ReceiverControl)
    case stop
}

public enum ReceiverControl: String, Codable, Sendable {
    case playPause
    case seekBack
    case seekForward
}

public struct ReceiverResponse: Codable, Sendable, Equatable {
    public let id: UUID
    public let payload: ResponsePayload

    public init(id: UUID, payload: ResponsePayload) {
        self.id = id
        self.payload = payload
    }
}

public enum ResponsePayload: Sendable, Equatable {
    case ok
    case answer(sdp: String)
    case error(message: String)
    // Mirrors the old 409: no active session for a control/stop to apply to.
    case notHandled
}

// Hand-written rather than relying on associated-value enum synthesis, so the
// wire shape (a `type` discriminator plus the relevant fields) is explicit,
// stable, and matches Receiver's decoding exactly.

extension RequestPayload: Codable {
    private enum CodingKeys: String, CodingKey { case type, payload, control }
    private enum Kind: String, Codable { case video, control, stop }

    public init(from decoder: Decoder) throws {
        let container = try decoder.container(keyedBy: CodingKeys.self)
        switch try container.decode(Kind.self, forKey: .type) {
        case .video:
            self = .video(payload: try container.decode(String.self, forKey: .payload))
        case .control:
            self = .control(try container.decode(ReceiverControl.self, forKey: .control))
        case .stop:
            self = .stop
        }
    }

    public func encode(to encoder: Encoder) throws {
        var container = encoder.container(keyedBy: CodingKeys.self)
        switch self {
        case .video(let payload):
            try container.encode(Kind.video, forKey: .type)
            try container.encode(payload, forKey: .payload)
        case .control(let control):
            try container.encode(Kind.control, forKey: .type)
            try container.encode(control, forKey: .control)
        case .stop:
            try container.encode(Kind.stop, forKey: .type)
        }
    }
}

extension ResponsePayload: Codable {
    private enum CodingKeys: String, CodingKey { case type, sdp, message }
    private enum Kind: String, Codable { case ok, answer, error, notHandled }

    public init(from decoder: Decoder) throws {
        let container = try decoder.container(keyedBy: CodingKeys.self)
        switch try container.decode(Kind.self, forKey: .type) {
        case .ok:
            self = .ok
        case .answer:
            self = .answer(sdp: try container.decode(String.self, forKey: .sdp))
        case .error:
            self = .error(message: try container.decode(String.self, forKey: .message))
        case .notHandled:
            self = .notHandled
        }
    }

    public func encode(to encoder: Encoder) throws {
        var container = encoder.container(keyedBy: CodingKeys.self)
        switch self {
        case .ok:
            try container.encode(Kind.ok, forKey: .type)
        case .answer(let sdp):
            try container.encode(Kind.answer, forKey: .type)
            try container.encode(sdp, forKey: .sdp)
        case .error(let message):
            try container.encode(Kind.error, forKey: .type)
            try container.encode(message, forKey: .message)
        case .notHandled:
            try container.encode(Kind.notHandled, forKey: .type)
        }
    }
}

File: src/lib/wireProtocol.ts
// Mirrors Sources/ReceiverProtocol/WireProtocol.swift's JSON envelope exactly:
// { id, payload: { type, ...fields } } in both directions, one request -> one
// response correlated by id. Kept as a hand-written discriminated union
// (rather than deriving types from schema) so the wire shape stays an
// explicit, obvious match for the Swift Codable implementation it talks to.

export type ReceiverControl = "playPause" | "seekBack" | "seekForward";

export type RequestPayload =
  // Raw share text (a bare URL, or freeform text with a URL embedded in
  // it, e.g. Dropout's "I'm watching X on Dropout\nhttps://...") — Screen's
  // VideoParserRegistry is responsible for recognizing the provider, this
  // side does no interpretation. Was `{ type: "youtube"; url }` back when
  // Screen only played YouTube; see jj change ruvsvwto in ../vga.
  | { type: "video"; payload: string }
  | { type: "offer"; sdp: string }
  | { type: "control"; control: ReceiverControl }
  | { type: "stop" }
  // Asks Screen for the TURN relay to use when building an SDP offer —
  // needed because Chrome's host ICE candidates are mDNS-obfuscated and
  // unresolvable by Screen's WebKit-based receiver.html; see TurnServer.swift.
  | { type: "iceConfig" };

export type ResponsePayload =
  | { type: "ok" }
  | { type: "answer"; sdp: string }
  | { type: "error"; message: string }
  // Mirrors the old 409: no active session for a control/stop to apply to.
  | { type: "notHandled" }
  // A "turn:host:port?transport=udp" URL, no credentials (LAN-only, no
  // auth — same trust model as the rest of this protocol).
  | { type: "iceConfig"; turnURL: string };

export interface ReceiverRequest {
  id: string;
  payload: RequestPayload;
}

export interface ReceiverResponse {
  id: string;
  payload: ResponsePayload;
}

export function encodeRequest(payload: RequestPayload): { id: string; json: string } {
  const id = crypto.randomUUID();
  const request: ReceiverRequest = { id, payload };
  return { id, json: JSON.stringify(request) };
}

// Swift's UUID JSON representation is uppercase; JS's crypto.randomUUID() is
// lowercase. Foundation's UUID(uuidString:) parses incoming requests
// case-insensitively either way, but it always *re-serializes* a response's
// id canonically uppercase regardless of the request's original casing — so
// the two are never a byte-for-byte match on the way back. This decode step
// only validates shape; ReceiverConnection.handleIncoming is what actually
// lowercases response.id before matching it against `pending`.
export function decodeResponse(data: string): ReceiverResponse | null {
  try {
    const parsed = JSON.parse(data);
    if (
      parsed &&
      typeof parsed.id === "string" &&
      parsed.payload &&
      typeof parsed.payload.type === "string"
    ) {
      return parsed as ReceiverResponse;
    }
  } catch {
    // fall through
  }
  return null;
}
